gtsfm/averaging/rotation/shonan.py

```diff
@@ -100,58 +112,75 @@ def _run_with_consecutive_ordering(
                 not be computed (either underconstrained system or ill-constrained system).
         """
 
-        obj = ShonanAveraging3(between_factors, self.__get_shonan_params())
+        logger.info(
+            "Running Shonan with %d constraints on %d nodes",
+            len(between_factors),
+            num_connected_nodes,
+        )
+        shonan = ShonanAveraging3(between_factors, self.__get_shonan_params())
 
-        initial = obj.initializeRandomly()
-        result_values, _ = obj.run(initial, self._p_min, self._p_max)
+        initial = shonan.initializeRandomly()
+        logger.info("Initial cost: %.5f", shonan.cost(initial))
+        result, _ = shonan.run(initial, self._p_min, self._p_max)
+        logger.info("Final cost: %.5f", shonan.cost(result))
 
         wRi_list_consecutive = [None] * num_connected_nodes
         for i in range(num_connected_nodes):
-            if result_values.exists(i):
-                wRi_list_consecutive[i] = result_values.atRot3(i)
+            if result.exists(i):
+                wRi_list_consecutive[i] = result.atRot3(i)
 
         return wRi_list_consecutive
 
-    def run(
+    def _nodes_with_edges(
+        self, i2Ri1_dict: Dict[Tuple[int, int], Optional[Rot3]], relative_pose_priors: Dict[Tuple[int, int], PosePrior]
+    ) -> Set[int]:
+        """Gets the nodes with edges which are to be modelled as between factors."""
```

Comment:
is this accurate? it looks like the function is finding a union over the edges in the two dicts..

Reply:
Yes. We cannot have cameras as inputs which do not have any edge. And the input to rotation averaging has just 1 connected components. So the measurement edges and the prior edges gives us all the nodes which can be estimated.

Reply:
okay, I got confused because of the between factor in the documentation. this function has nothing do with between factors. You could also pass in .keys() of two dicts, and add them to a set here, call it _union_of_edges or something. But I don't have a preference, the current implementation is fine.

Reply:
I call it _nodes_with_edges because we return nodes and not edges.